### ISSUE TITLE
[ReactNavigationInstrumentation] Will not send transactiondue to beforeNavigate on react navigation v5

### ISSUE THREAD

**Issue by Lerosen** [getsentry/sentry-react-native#4109]
**OS:**
- [x]  Windows
- [ ]  MacOS
- [ ]  Linux

**Platform:**
- [ ]  iOS
- [X]  Android

**SDK:**
- [X]  `@sentry/react-native` (>= 1.0.0)
- [ ]  `react-native-sentry` (<= 0.43.2)

**SDK version:** 5.32.0

**`react-native` version:** 0.74.3

**Are you using Expo?**
- [X]  Yes
- [ ]  No

**Are you using sentry.io or on-premise?**
- [X]  sentry.io (SaaS)
- [ ]  on-premise

**If you are using sentry.io, please post a link to your issue so we can take a look:**

[Link to issue]

**Configuration:**

(`@sentry/react-native`)
```
const routingInstrumentation = Sentry.reactNavigationIntegration({
  enableTimeToInitialDisplay: true
});

Sentry.init({
  dsn: "XXX",
  debug: true, // If `true`, Sentry will try to print out useful debugging information if something goes wrong with sending the event. Set it to `false` in production
  enabled: !__DEV__ || true,
  environment: __DEV__ ? "development" : "production",
  integrations: [
    Sentry.reactNativeTracingIntegration({
      routingInstrumentation
    })
  ],
  enableTracing: true,
  tracesSampleRate: 1.0
});
```

---
**I have the following issue:**

I am using Sentry with ReactNavigation v5 ("@react-navigation/native": "^5.9.8").
I set up the screen tracing instrumentation as documented. For debugging purposes I also added my own logging of routing events.

The issue that I don't get any Screen load events. THe log indicated that all navigation events are correctly detected but each transaction is discarded because the sample property is false. I tried to find the cause for this but had no luck so far. I don't have any beforeNavigate handler configured.

**Steps to reproduce:**
- Navigate to some screen, observe the following log

**Actual result:**

```
  Sentry Logger [log]: [TouchEvents] Touch event within element: View
 LOG  Sentry Logger [log]: [TouchEvents] Touch event within element: View
 LOG  Sentry Logger [log]: Setting idle transaction on scope. Span ID: 846bcc77673deb37
 DEBUG  Sentry Logger [debug]: [NativeFrames] Fetching frames for root span start (846bcc77673deb37).
 LOG  Sentry Logger [log]: [ReactNativeTracing] Starting navigation transaction "Route Change" on scope
 LOG  before[event=__unsafe_action__] undefined {"action": {"payload": {"name": "test_screen", "params": undefined}, "target": "tab-eoC-AJVrF_9H8QZchBZHd", "type": "JUMP_TO"}, "noop": false}
 LOG  Sentry Logger [log]: [ReactNavigationInstrumentation] Will not send transaction "test_screen" due to beforeNavigate.
 LOG  after[event=state] {"key": "update_notes-BxXRVcpR-K59Cvc6hnoxB", "name": "test_screen", "params": undefined}
 LOG  Sentry Logger [log]: [Tracing] No active IdleTransaction
 LOG  Sentry Logger [log]: [Tracing] Discarding transaction because its trace was not chosen to be sampled.
 LOG  Sentry Logger [log]: Recording outcome: "sample_rate:transaction"
```

**Expected result:**
The transaction should not be discarded but send.

**Comment by rwachtler** [comment]
Hey @Lerosen! 👋 

Since I'm not a seasoned contributor, please consider my advice as my personal take at your issue. I've compared your configuration to the [documented reference](https://docs.sentry.io/platforms/react-native/tracing/instrumentation/react-navigation/#initialization) (and also cross-checked the examples [here](https://github.com/getsentry/sentry-react-native/pull/4073), since you're using `5.32.0`). Didn't really discover any discrepancies there and given the logs you provided, the instrumentation is obviously detecting the events.

**TLDR:** I assume that you run into a [state-change timeout](https://github.com/getsentry/sentry-react-native/blob/a9cf86cb6403e41c8ab8e47c63462210cd32b107/src/js/tracing/reactnavigation.ts#L212C10-L212C29). You could try setting [`routeChangeTimeoutMs`](https://docs.sentry.io/#routechangetimeoutms) to some higher value than `1000`.

**Long version:**

Here is my shot at what's happening:

1. I've looked up the log 

```bash
[ReactNavigationInstrumentation] Will not send transaction [...] due to beforeNavigate.
```

and found it [here]([ReactNavigationInstrumentation] Will not send transaction "test_screen" due to beforeNavigate.), which gave me a clue that `finalContext.sampled` must be still `false`.

2. Since you're not manipulating the `sampled` prop, I've also searched for `__unsafe_action__` which led me to `_onDispatch` [here](https://github.com/getsentry/sentry-react-native/blob/a9cf86cb6403e41c8ab8e47c63462210cd32b107/src/js/tracing/reactnavigation.ts#L191).

It's description being the following:

```
/**
   * To be called on every React-Navigation action dispatch.
   * It does not name the transaction or populate it with route information. Instead, it waits for the state to fully change
   * and gets the route information from there, @see _onStateChange
   */
```

the description above plus an explanation 

3. While looking into the body of `_onDispatch` , I was wondering if you actually run into a [state-change timeout](https://github.com/getsentry/sentry-react-native/blob/a9cf86cb6403e41c8ab8e47c63462210cd32b107/src/js/tracing/reactnavigation.ts#L212C10-L212C29). As noted [here](https://github.com/getsentry/sentry-react-native/blob/a9cf86cb6403e41c8ab8e47c63462210cd32b107/src/js/tracing/reactnavigation.ts#L63)

```
 * - If `_onStateChange` isn't called within `STATE_CHANGE_TIMEOUT_DURATION` of the dispatch, then the transaction is not sampled and finished.

```

resulting in `sampled` being set to `false` (see [this](https://github.com/getsentry/sentry-react-native/blob/a9cf86cb6403e41c8ab8e47c63462210cd32b107/src/js/tracing/reactnavigation.ts#L365)).

4. Resolution proposal would be setting [`routeChangeTimeoutMs`](https://docs.sentry.io/#routechangetimeoutms) > `1000`.

**Comment by Lerosen** [comment]
Thank you for taking the time to look into this and for your detailed answer. But the timeout is not the problem, it is triggered much later, way after the log  `iscarding transaction because its trace was not chosen to be sampled.`.
I also tried a higher value of `10000` for the `routeChangeTimeoutMs`, but that didn't help.

**Comment by rwachtler** [comment]
Thank you for providing additional context!  

> But the timeout is not the problem, it is triggered much later, way after the log iscarding transaction because its trace was not chosen to be sampled..

Hmm, good point. I'm currently on my mobile, I will look into it further once I've access to my Mac again. 😅

**Comment by lucas-zimerman** [comment]
Hi and thank you for opening this issue!
Did the transaction get discarded right away, or after the defined time on `routeChangeTimeoutMs`?

**Comment by Lerosen** [comment]
> Hi and thank you for opening this issue!
> Did the transaction get discarded right away, or after the defined time on `routeChangeTimeoutMs`? 

As I said, it is discarded right away. Sry, I missed your reply.

**Comment by krystofwoldrich** [comment]
Hi @Lerosen,
thank you for the message, I've check the logs and previous answers to the issue and it looks like that tracing might be evaluated as disabled when checked during the transaction creation.

https://github.com/getsentry/sentry-javascript/blob/f58bf69cfa771ba5d5ef65c7ff8bb8c103264d0a/packages/core/src/tracing/sampling.ts#L27

From the logs sadly I don't see why this would happen.

Would you be able to debug the linked code or share a reproducible sample app with us, so we can debug this?

**Comment by Lerosen** [comment]
I started setting up a reproduction in an expo snack but had no time to finish it yet.

**Comment by krystofwoldrich** [comment]
Closing this due to inactivity. If this is still an issue feel free to comment here or open a new issue if the context changed.
